[Patterns]: Browse Patterns in a Modal

ntsekouras opened the pull request:
This PR is another experiment for a better pattern insertion experience with the related design: https://github.com/WordPress/gutenberg/pull/35006#issuecomment-929605688.

Related issues (https://github.com/WordPress/gutenberg/issues/26905, https://github.com/WordPress/gutenberg/issues/35274, https://github.com/WordPress/gutenberg/issues/31153)
Different pattern explorer experiment: https://github.com/WordPress/gutenberg/pull/35006




## Show case

https://user-images.githubusercontent.com/16275880/138070824-802ea222-b1b7-4924-8415-d3e6437a905e.mov




## Notes
Clicking a pattern in the modal (if you have search anything) will not close the modal and is known.

ntsekouras committed:
[Patterns]: Paterns explorer in modal

ntsekouras committed:
add search results header info

mcsf commented on packages/block-editor/src/components/inserter/block-patterns-explorer/search-results.js:
```diff
@@ -0,0 +1,103 @@
+						),
+						filteredBlockPatterns.length
+					) }
+					{ ` "${ filterValue }"` }
```
`filterValue` should be a part of the translator string:

```suggestion
					{ sprintf(
						/* translators: %d: number of patterns. %s: block pattern search query */
						_n(
							'%d pattern found for "%s"',
							'%d patterns found for "%s"',
							filteredBlockPatterns.length
						),
						filteredBlockPatterns.length,
						filterValue
					) }
```

mcsf reviewed (commented):
This is looking good, but I can definitely feel the rendering locking up my process for a few seconds as I open or close the explorer (Safari, dev build). Wondering what we can do there.

ntsekouras committed:
Update packages/block-editor/src/components/inserter/block-patterns-explorer/search-results.js

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

ntsekouras commented:
>but I can definitely feel the rendering locking up my process for a few seconds as I open or close the explorer (Safari, dev build). Wondering what we can do there.

I know :)  I made the PR to gather the design wise feedback first. I'll have to check the performance and make adjustments like moving state in lower levels etc..

paaljoachim commented:
This looks very interesting!

This is what I see.

![Screenshot 2021-10-20 at 22 01 50](https://user-images.githubusercontent.com/5323259/138163536-5a7dd14a-f094-4a8f-be87-1250bf643658.jpg)

Tool tip says "Pattern explorer". Should it instead say "Explore patterns" ?

Clicking the extender (what I call it) I see the Pattern explorer. Gallery of patterns.

![Screenshot 2021-10-20 at 22 03 32](https://user-images.githubusercontent.com/5323259/138163771-851e003e-d505-4d88-9d42-c84e735328d1.jpg)

What if we instead of having a black button for selected area just use an underline (and overline).
Here is an example of underline and overline. There is less contrast between the black button with the white text and the surrounding black text.
![Pattern-explorer-under-overline-selected-category](https://user-images.githubusercontent.com/5323259/138164480-d5fdf127-0bee-4b91-b2fb-42a9e9453d9f.jpg)
(I assume we should look at how WordPress uses color for selected links.)

Sidebar links should be sticky/fixed in place so that scrolling will keep the links visible in the sidebar. So one would not scroll a long page of patterns and need to scroll up to change to another pattern category.

I assume when clicking to add a pattern that the extender would automatically close (and Block inserter would close) so that we can see the new pattern added to the canvas.

It is a very nice beginning Nik! Good work!

annezazu commented:
Just tested! Excited to see this coming together. Personally, I don't find the icon used to open this up to be the right one (not sure if design came up with this though but wanted to note) - it feels very easy to not find or use as a default experience. 

As for the UX, it's _mostly_ smooth but I found lots of double scrollbars that made the experience very janky:

https://user-images.githubusercontent.com/26996883/138311063-173ad868-ab2e-4f52-bf86-e763cd90b0d4.mov

shaunandrews commented:
Very cool. In addition to the scrollbar issue Anne mentions, I think there's a few other things to do to make this ready:

--

I think the Fullscreen icon probably isn't the best way to expose this modal. Instead, a button is probably the best option.

<img width="444" alt="image" src="https://user-images.githubusercontent.com/191598/138334355-14a77efb-aa70-4fc6-9ba7-cdb08ed8bb72.png">

--

The modal's sidebar shouldn't scroll with the rest of the modal.

--

The menu in the sidebar should stretch to the same width as the search. I think it would be nice to add a Newest item, and group it with Featured. Then the categories would be grouped under that.

<img width="282" alt="image" src="https://user-images.githubusercontent.com/191598/138335095-4e9eb071-fff9-47e0-8c96-3ea69b0155e2.png">

If it helps, here's a redline from the design file for the spacing around the sidebar menu. My designs have the sidebar at 240px, with the grid of patterns taking up the remainder of the space.

<img width="667" alt="image" src="https://user-images.githubusercontent.com/191598/138336085-63042702-532a-4212-8685-44f30741b7a6.png">

--

Here's a design for the modal showing a section title, a "sort by" dropdown menu, the grid of results, and a pager:

<img width="1384" alt="image" src="https://user-images.githubusercontent.com/191598/138336278-9a7d808f-2e0a-4a51-974b-08ddbfcacd2b.png">

The "sort by" dropdown probably isn't necessary, though I imagine it could be interesting to sort by newest first, oldest first, most favorites, insertions, and more.

--

It would be nice to add a light-gray border around the pattern previews (not including the pattern name). When hovering/focusing a pattern in the grid, the border should turn blue (primary color) and an overlay+button could appear to help explain what happens when you click:

<img width="731" alt="image" src="https://user-images.githubusercontent.com/191598/138337181-0aca25ef-5878-4f03-9785-c0742fd2260f.png">

In a follow up PR it might be nice to add a second button to this hover to "Learn more" about a pattern. Clicking the learn more button would display the pattern detail page in the modal; This isn't necessary for this first pass of the feature.

shaunandrews commented:
Here's a quick video of how the above designs come together:

https://user-images.githubusercontent.com/191598/138357481-b156f121-9bd8-4b31-b22a-6b22aa6d6780.mp4

ntsekouras commented:
Thanks for the feedback! I'll work on some updates.

Just noting though that the concept of `sort` and newest is not applicable yet, as the patterns don't have a date field and are registered/shown by their registration order. Same goes with categories with an exception of `featured` and `uncategorized` that are shown first and last respectively.

ntsekouras committed:
fix lint error

ntsekouras committed:
small design changes

jameskoster commented:
I think we need an outline on patterns in their resting state too. Otherwise the grid feels poorly defined when there are lots of patterns with white backgrounds:

<img width="1508" alt="Screenshot 2021-10-22 at 09 43 03" src="https://user-images.githubusercontent.com/846565/138423149-f65053b8-3f46-47f6-a9b9-198a65974f11.png">

shaunandrews commented:
Yup the design above suggests a border for the resting state.

ntsekouras committed:
Refactor patterns tab/explorer for performance

jasmussen commented:
Took this one for a spin. It's come a remarkable way in a short time:
![patterns modal](https://user-images.githubusercontent.com/1204802/138675628-c8bfe7f3-a28e-4a5f-bf88-17ef062cba64.gif)

Overall this does give a bigger space to explore patterns in, if you're just looking at what's available. I think the downsides to a modal pattern dialog remain: 

- It's entirely disconnected from the left sidebar interface (did we have a mockup where it expands that one instead?)
- It covers the entire editing canvas obscuring the spot at which you'll insert the pattern
- Outside of being literally disconnected, it's visually different from the sidebar interface, making it feel all the more separate. Border around patterns in one but not the other, for example, if we were to ever make the expanding sidebar, these would need to be the same.

I wouldn't personally object to a "merge and iterate" strategy on this one. I do like the idea of offering a large canvas for patterns directly in core. But I do think iterations in the future can improve the connective tissue between sidebar and pattern drawer.

I noticed some glitches in the GIF above, including that the individual pattern thumbnails scroll in some cases. Seems like we need an overflow: hidden rule. Much of the rest has already been mentioned. 

Beautiful mockups have already been shared, but I do wonder: instead of introducing a new left hand tabbar component with inverted selected state, should we reuse the one that's being used for the Preferences dialog?
<img width="261" alt="Screenshot 2021-10-25 at 12 11 27" src="https://user-images.githubusercontent.com/1204802/138677593-6711703b-1c0f-4462-b919-e58e3ef3357a.png">

I don't mean to object to the new metrics or visuals of the refresh, just wondering if we should use the same component and unify the design?

ntsekouras commented:
>It's entirely disconnected from the left sidebar interface (did we have a mockup where it expands that one instead?)

I think @shaunandrews had some designs about this - maybe [this one](https://github.com/WordPress/gutenberg/pull/35006#issuecomment-925097765)?

>I noticed some glitches in the GIF above, including that the individual pattern thumbnails scroll in some cases. Seems like we need an overflow: hidden rule. Much of the rest has already been mentioned.

It's not a glitch, just a `max-height` I added in this PR which can be removed 😄

jasmussen commented:
> It's not a glitch, just a max-height I added in this PR which can be removed 😄

It's fine to keep the max-height and "soft crop" the thumbnails if we need to. They just shouldn't be scrollable also 🚀 

And I liked the [expanding sidebar](https://github.com/WordPress/gutenberg/pull/35006#issuecomment-925097765)!

shaunandrews commented:
> It's entirely disconnected from the left sidebar interface (did we have a mockup where it expands that one instead?)

I'm not sure what you mean here. To me, the modal feels like a natural next step for adding patterns.

<img width="1828" alt="image" src="https://user-images.githubusercontent.com/191598/138728706-abc5893b-98c6-4695-be68-65f33849210b.png">

I think we could also connect all of these surfaces together with motion:

https://user-images.githubusercontent.com/191598/138729893-cd17180c-ae59-4be6-81ca-a70199dd5790.mp4

> It covers the entire editing canvas obscuring the spot at which you'll insert the pattern

It does, but I don't think that's a problem; There are still the sidebar and quick inserter, and I believe there are many improvements we could make to those experiences to help emphasize "the spot which you'll insert." Here's an idea from a few weeks ago for showing a preview when adding a new pattern from the sidebar inserter:

https://user-images.githubusercontent.com/191598/138729303-369e2e02-a58c-4c58-96ee-8fca8e5feef1.mp4

> Outside of being literally disconnected, it's visually different from the sidebar interface, making it feel all the more separate. Border around patterns in one but not the other, for example, if we were to ever make the expanding sidebar, these would need to be the same.

Other than the border around patterns (which I strongly suggest we _do_ add everywhere) I'm not sure what other visual differences you find concerning?

--

> I don't mean to object to the new metrics or visuals of the refresh, just wondering if we should use the same component and unify the design?

I did start with that selected state for the sidebar items, but it conflicts with the search input:

<img width="389" alt="image" src="https://user-images.githubusercontent.com/191598/138730405-09ec36f8-a915-4e8d-ba9c-e9a45232f555.png">

I have looked at a number of other options, but nothing seemed as good as the inverted select state:

<img width="1760" alt="image" src="https://user-images.githubusercontent.com/191598/138732608-bb90a0fd-c2db-4118-89b1-67b7dc4977ee.png">

jasmussen commented:
To clarify, I think it's important we offer an interface like this so plugins don't have to provide their own. I dig your work, and I hope to help Nik land this one with some CSS polish if we can get it to a good place in the near term. In the longer term, I do imagine there are some general flows that could be improved, so that's more of a question of how/if we can iterate this approach.

> I'm not sure what you mean here. To me, the modal feels like a natural next step for adding patterns.

I meant in the literal sense. It's a new interface that opens from another, like an inserter inside the inserter. It can work, but it's for this reason I had a preference for your [expanding sidebar](https://github.com/WordPress/gutenberg/pull/35006#issuecomment-925097765) instead.

> It does, but I don't think that's a problem; There are still the sidebar and quick inserter, and I believe there are many improvements we could make to those experiences to help emphasize "the spot which you'll insert." Here's an idea from a few weeks ago for showing a preview when adding a new pattern from the sidebar inserter:

Agreed, and excellent animatic.

And if we accept the tradeoff of covering content, we can offer something else instead. In this case I feel like we should at least show the patterns larger than we do them in the sidebar. Perhaps just showing 2 columns instead of 3 columns of patterns would help here. 

> Other than the border around patterns (which I strongly suggest we do add everywhere) I'm not sure what other visual differences you find concerning?

In your animatic above, I liked the icon for opening. The black button looks stylish, but it stands out a little in a way that feels unbalanced. In absence of going back to an icon, it also seems like we could find a better label than "See all", as you can see them all in the sidebar interface. "Explore" might not be a great alternative — but something that indicates it's more of an immersive interface.

> I did start with that selected state for the sidebar items, but it conflicts with the search input:

I like the gray for managing the prominence of the active tab. For a brief moment I wondered if we should use a bordered version of search here instead of the gray solid one, but that would introduce its own discrepancy. 

More importantly than the color of the tab, it just seems like we should be using the same component as is used in the Preferences modal, and update the component itself if it needs to change colors, so that it will be the same in both places. 

How would the tabs look if bordered? I saw some "style variation" buttons recently that were bordered and looked good, would that style work here?

An alternative is to put the search somewhere else, in the header or on the right:

<img width="777" alt="Screenshot 2021-10-25 at 18 58 52" src="https://user-images.githubusercontent.com/1204802/138738568-6c89caa5-e1fb-458a-8cc3-fcf44df22e03.png">

There's a bunch of reasons that might not work, and the dark tabs might work fine in preferences as well!

ntsekouras commented:
Do you think we can decide on the remaining design issues? If this is something we are not sure we want for 5.9, then it's okay to take our time. If it's something we want, we should land a the first version and iterate later 😄

shaunandrews commented:
I spent the day today tweaking some of the visuals on this PR and pushed some (mostly) CSS tweaks. Here's how it looks right now:

https://user-images.githubusercontent.com/191598/139740008-5534fc64-570a-4a45-a16e-8910b889bc8a.mp4

The one thing I couldn't get working was the fixed positioning of the modal's sidebar; I don't expect that it should scroll along with the list of patterns. Otherwise, I think this is a great start and would love to see it merged and iterated on for 5.9.

shaunandrews committed:
Moving hover and focus styles to the preview container.

shaunandrews committed:
Specifying a gap for the Grid component.

shaunandrews committed:
Updating the button label to Explore and using the outlined style of button.

shaunandrews committed:
Visual tweaks to the inserter modal

- No need for custom styles on the explore button.
- No need to separate the Featured tab on its own.
- Switching from 3 to 2 columns for the list.
- Increased the max-height and ensured no scrolling within the preview container.

ntsekouras committed:
adjust min-height in quick inserter

ntsekouras commented on packages/block-editor/src/components/inserter/block-patterns-explorer/patterns-list.js:
```diff
@@ -0,0 +1,126 @@
+		return null;
+	}
+	return (
+		<h2 className="block-editor-block-patterns-explorer__search-results-count">
```
This has a class now but no styles for it and will take the current theme's styles. We should probably add some styles here, no?

shaunandrews replied:
I switched this out to use the Heading component.

ntsekouras replied:
Thanks so much!

ntsekouras committed:
remove commented line

ntsekouras commented:
Thanks for the polishing here @shaunandrews ! 💯

shaunandrews committed:
Adjusting min/max height for previews in the modal and changing columns at breakpoints.

shaunandrews committed:
More visual tweaks to the patterns list.

- Moving the margin to the bottom and increasing it to give the list some room to breathe.
- Removing an unnecessary position, background, and transition.
- Adjusting the title's padding to only apply to the top and adding a hover state.

shaunandrews committed:
Increasing the space between the dropdown and the button to match the other spacing in the sidebar.

shaunandrews committed:
Using the Heading component and adding a className to the list container div.

shaunandrews committed:
Absolute positioning the explorer sidebar to allow for independent scrolling.

shaunandrews commented:
I pushed a few more changes, including absolute positioning for the modal sidebar to allow for independent scrolling. I also added some breakpoints for the main grid to allow for 2 or 3 columns depending.

I wonder: Should we hide the Explore button from the sidebar on mobile? I'm not sure if the modal makes much sense on modal, as its essentially the same width as the sidebar.

jasmussen commented:
> I wonder: Should we hide the Explore button from the sidebar on mobile? I'm not sure if the modal makes much sense on modal, as its essentially the same width as the sidebar.

Yeah that makes sense. The larger canvas is intrinsically a tablet/desktop experience.

ntsekouras committed:
hide explore button when inserter goes full width + small css fix for pattern item in quick inserter results

ntsekouras commented:
>I wonder: Should we hide the Explore button from the sidebar on mobile?

I updated to hide the explore button when inserter goes full width. Thanks again for the help here @shaunandrews and @jasmussen! 🥇

critterverse commented:
Wow, this feels so clean and refined. Great work all.

I tested this in a few different contexts and it looks good all around. One small note that the patterns grid appears to bump up against the left column when scroll bars are visible by default (mac OS setting):

<img width="1087" alt="always-on-scrollbars" src="https://user-images.githubusercontent.com/77359364/140431437-39fdc62b-c6af-4b78-9d41-26e13b6631dd.png">

mcsf commented on packages/components/src/modal/index.js:
```diff
@@ -52,6 +52,7 @@ export default function Modal( {
 	className,
 	contentLabel,
 	onKeyDown,
+	isFullScreen, // TODO: change docs
```
Now is a good reminder to address this. :)

ntsekouras replied:
😆

mcsf commented on packages/block-editor/src/components/inserter/block-patterns-explorer/patterns-list.js:
```diff
@@ -0,0 +1,130 @@
+	filterValue,
+	selectedCategory,
+	patternCategories,
+	maxSearchResults,
```
Looks like this prop isn't used anywhere. Shall we remove it? [YAGNI]

mcsf commented on packages/block-editor/src/components/inserter/block-patterns-explorer/patterns-list.js:
```diff
@@ -0,0 +1,130 @@
+	return (
+		<Heading
+			level={ 2 }
+			lineHeight={ '48px' }
```
Minor note, but what could we do instead of keeping this magic constant?

shaunandrews replied:
I expect this would go away as we iterate more; Right now it helps resolve a visual imbalance when searching, but I expect the headings to change (like adding descriptions or actions) as we add them to the other parts of the browsing experience.

mcsf reviewed (approved):
Nice work, really polished!

paaljoachim reviewed (approved):
From a UI and UX view point this looks really nice!

https://user-images.githubusercontent.com/5323259/140552900-05bb4fba-45a9-4eb3-a633-b5ee2a23fd75.mp4

ntsekouras committed:
address feedback

andrewserong commented:
Cherry-picked into the GB 11.9 release in https://github.com/WordPress/gutenberg/commit/8c598f4d548f72531c01a7cec5b120e56842461e

noisysocks commented:
This made it into WP 5.9 because it landed in `release/11.9` before I published the initial packages.
